README
MQCrawler is a distributed perl plateform for fast crawling web sites



APACHE ACTIVEMQ SERVER
-----------------------

You need an activemq server with "stomp" protocol activated
you also need to specify your activemq server address
into YSpider.conf


REDIS SERVER
-----------------------

A Redis server is needed in order to store already crawled url
(into the "url" set)
with Ubuntu, you can "apt-get install redis-server"
and specify your redis server address into YSpider.conf

PERL PROGRAMS
-----------------------

please, specify as many patterns as you need
in YSpider.conf


then, you have to start in this order :

> perl master.pl &
> perl analyzer.pl &
> perl crawler.pl &


master.pl   :
------------------
 retrieves the links (from the "links" queue) to crawl and send the autorized urls to be crawled to the "crawl" queue   

analyzer.pl : 
------------------
 retrieves the html source (from the "source" topic) and extracts the links present into this source and send them to the "links" queue 

 you can write as many analyzer.pl  "clones" for your own purposes, as source code are published into the "source" topic, with persistent message ;)

crawler.pl  : 
------------------

 retrieves the urls to crawl from the "crawl" queue and gets the source content and send them to the "source" topic.
 this script forks n times, in order to parallelize http gets.

 you can launch crawler.pl on as many servers as you want/need.

 be carefull, it may DDOS the web sites you wanna crawl.


when everything is done, you can send the first url to crawl to the "crawl" queue with a starting url in the body message.


CHANGELOG
2012-02-24

send into "source" topic/queue http-headers, status-code, redirects


2012-02-19

some improvements for configuration files
and change some variables, some values


master.pl
#!/usr/bin/perl 
#
#
use strict;
use warnings;
use Net::Stomp;
use Data::Dump qw(dump);
use HTML::Parser;
use JSON;
use AppConfig;
use AppConfig::State;
use AppConfig::File;
use POSIX;
use File::Pid;
use Redis;

my $daemonName = "MQmaster";

#=======================================
# Lecture du fichier de config
#=======================================
my $config = AppConfig->new();
   $config->define("patterns",      {ARGCOUNT => AppConfig::ARGCOUNT_LIST});
   $config->define("redis_server",  {ARGCOUNT => AppConfig::ARGCOUNT_ONE});
   $config->define("redis_set_urls",  {ARGCOUNT => AppConfig::ARGCOUNT_ONE});
   $config->define("redis_set_urls_crawled",  {ARGCOUNT => AppConfig::ARGCOUNT_ONE});
   $config->define("activemq_server",  {ARGCOUNT => AppConfig::ARGCOUNT_ONE});
   $config->define("activemq_source",  {ARGCOUNT => AppConfig::ARGCOUNT_ONE});
   $config->define("activemq_links",  {ARGCOUNT => AppConfig::ARGCOUNT_ONE});
   $config->define("activemq_crawl",  {ARGCOUNT => AppConfig::ARGCOUNT_ONE});
   $config->define("logging",       {ARGCOUNT => AppConfig::ARGCOUNT_ONE});
   $config->define("logpath",       {ARGCOUNT => AppConfig::ARGCOUNT_ONE});
   $config->define("pidpath",       {ARGCOUNT => AppConfig::ARGCOUNT_ONE});
   $config->file('YSpider.conf');

   my @patterns_to_crawl   = ();
   my $patterns = $config->patterns;
   my $i;
   foreach $i (@$patterns) {
                push( @patterns_to_crawl, $i);
   }


   my $redis_server  = $config->redis_server;
   my $redis_set_urls = $config->redis_set_urls;
   my $redis_set_urls_crawled = $config->redis_set_urls_crawled;

   my $activemq_server  = $config->activemq_server;
   my $activemq_source  = $config->activemq_source;
   my $activemq_links   = $config->activemq_links;
   my $activemq_crawl   = $config->activemq_crawl;

   my $logging       = $config->logging();                           # 1= logging is on
   my $logFilePath   = $config->logpath();                           # log file path
   my $pidFilePath   = $config->pidpath();                           # PID file path

   $logFilePath   =~ s/([^\/])$/$1\//; # add slash / if not specified in config 
   $pidFilePath   =~ s/([^\/])$/$1\//; # add slash / if not specified in config

   my $logFile       = $logFilePath . $daemonName . ".log";
   my $pidFile       = $pidFilePath . $daemonName . ".pid";

sub GiveMeNextLinksToCrawl
{
    eval {
	my $link;
	my @taboflinks;
 	my $stomp = Net::Stomp->new( { hostname => $activemq_server, port => '61613'} );

	$stomp->connect();
	$stomp->subscribe(
        {   destination             => $activemq_links,
                'ack'                   => 'client',
                'activemq.prefetchSize' => 1
        });

	my $r = Redis->new( server => $redis_server);

	while($stomp->can_read({ timeout => '1' }) eq 1) {
	    	my $frame = $stomp->receive_frame;
    		my $json_links= $frame->body; # do something here
		my $url   = $frame->headers->{"correlation-id"};
		my $ok    = $r->sadd($redis_set_urls_crawled, $url );

		logEntry("pop ".$url);
		@taboflinks=();
		@taboflinks = @{decode_json($json_links)};
		foreach $link (@taboflinks) {
                        if (URL_isAutorised( $link->{"href"} ) eq "true" )
                        {
                                if (! $r->sismember( $redis_set_urls,  $link->{"href"} ))
				{
                                        # print $link->{"href"}."\n";
                                        logEntry("push ".$link->{"href"});
                                        my $frame_producer = Net::Stomp::Frame->new( {
                                                body    => $link->{"href"},
                                                command => "SEND",
                                                headers => {
                                                        "destination"    => $activemq_crawl,
                                                        "correlation-id" => $link->{"href"},
                                                        "persistent"     => 'true'
                                                }
                                        });
                                        $stomp->send_frame($frame_producer);
                                        $ok = $r->sadd($redis_set_urls, $link->{"href"} );
                                }
                        }
                }
		$stomp->ack( { frame => $frame } );
	}
	$r->quit;
  	$stomp->disconnect;
  };
  logEntry($@) if $@;
  return;
}

sub URL_isAutorised {           # Verify if the given url can be crawled by configuration file (patterns) or robots.txt
        my ($url) = @_;
        my ($pattern);

        foreach $pattern (@patterns_to_crawl) 
	{
                if (substr($url,0,length($pattern)) eq $pattern)
	        {
                        return "true";
                }
        }
        return "false";
}
 
#################################################################################################
# MAIN PROGRAM
#################################################################################################

       my $dieNow        = 0;                                     # used for "infinte loop" construct - allows daemon mode to gracefully exit
       my $sleepMainLoop = 10;                                    # number of seconds to wait between "do something" execution after queue is clear

       # daemonize
       use POSIX qw(setsid);
       chdir '/';
       umask 0;
       open STDIN,  '/dev/null'   or die "Can't read /dev/null: $!";
       open STDOUT, '>>/dev/null' or die "Can't write to /dev/null: $!";
       open STDERR, '>>/dev/null' or die "Can't write to /dev/null: $!";

       # dissociate this process from the controlling terminal that started it and stop being part
       # of whatever process group this process was a part of.
       POSIX::setsid() or die "Can't start a new session.";

       # callback signal handler for signals.
       $SIG{INT} = $SIG{TERM} = $SIG{HUP} = \&signalHandler;
       $SIG{PIPE} = 'ignore';

       # create pid file in /var/run/
       my $pidfile = File::Pid->new( { file => $pidFile, } );

       $pidfile->write or die "Can't write PID file, /dev/null: $!";

       # turn on logging
       if ($logging) {
            open LOG, ">>$logFile";
            select((select(LOG), $|=1)[0]); # make the log file "hot" - turn off buffering
       }

       logEntry("daemon started");

       # Enter loop to do work
       # "infinite" loop where some useful process happens
       until ($dieNow) {
          GiveMeNextLinksToCrawl(); 
  	  sleep($sleepMainLoop);
       }

exit;

# add a line to the log file
sub logEntry {
    my ($logText) = @_;
    my ( $sec, $min, $hour, $mday, $mon, $year, $wday, $yday, $isdst ) = localtime(time);
    my $dateTime = sprintf "%4d-%02d-%02d %02d:%02d:%02d", $year + 1900, $mon + 1, $mday, $hour, $min, $sec;
    if ($logging) {
        print LOG "$dateTime|$daemonName|$logText\n";
    }
}

# catch signals and end the program if one is caught.
sub signalHandler {
    $dieNow = 1;    # this will cause the "infinite loop" to exit
}

# do this stuff when exit() is called.
END {
    logEntry("daemon stopped");
    if ($logging) { close LOG }
    $pidfile->remove if defined $pidfile;
}


analyzer.pl
#!/usr/bin/perl 
#
use strict;
use warnings;
use Net::Stomp;
use Data::Dump qw(dump);
use HTML::Parser;
use JSON;
use URI::URL;
use Encode;
use POSIX;
use File::Pid;
use AppConfig;

my $daemonName = "MQanalyzer";

#=======================================
# Lecture du fichier de config
#=======================================
my $config = AppConfig->new();
   $config->define("patterns",      {ARGCOUNT => AppConfig::ARGCOUNT_LIST});
   $config->define("redis_server",  {ARGCOUNT => AppConfig::ARGCOUNT_ONE});
   $config->define("redis_set_urls",  {ARGCOUNT => AppConfig::ARGCOUNT_ONE});
   $config->define("redis_set_urls_crawled",  {ARGCOUNT => AppConfig::ARGCOUNT_ONE});
   $config->define("activemq_server",  {ARGCOUNT => AppConfig::ARGCOUNT_ONE});
   $config->define("activemq_source",  {ARGCOUNT => AppConfig::ARGCOUNT_ONE});
   $config->define("activemq_links",  {ARGCOUNT => AppConfig::ARGCOUNT_ONE});
   $config->define("activemq_crawl",  {ARGCOUNT => AppConfig::ARGCOUNT_ONE});
   $config->define("logging",       {ARGCOUNT => AppConfig::ARGCOUNT_ONE});
   $config->define("logpath",       {ARGCOUNT => AppConfig::ARGCOUNT_ONE});
   $config->define("pidpath",       {ARGCOUNT => AppConfig::ARGCOUNT_ONE});
   $config->file('YSpider.conf');

   my @patterns_to_crawl   = ();
   my $patterns = $config->patterns;
   my $i;
   foreach $i (@$patterns) {
                push( @patterns_to_crawl, $i);
   }


   my $redis_server  = $config->redis_server;
   my $activemq_server  = $config->activemq_server;
   my $activemq_source  = $config->activemq_source;
   my $activemq_links   = $config->activemq_links;
   my $activemq_crawl   = $config->activemq_crawl;

   my $logging       = $config->logging();                           # 1= logging is on
   my $logFilePath   = $config->logpath();                           # log file path
   my $pidFilePath   = $config->pidpath();                           # PID file path

   $logFilePath   =~ s/([^\/])$/$1\//; # add slash / if not specified in config 
   $pidFilePath   =~ s/([^\/])$/$1\//; # add slash / if not specified in config

   my $logFile       = $logFilePath . $daemonName . ".log";
   my $pidFile       = $pidFilePath . $daemonName . ".pid";


my @taboflinks=();

sub parseur_start {
        my ($tag, $attr,$tableau) = @_;
		$tag=lc($tag);
                if ($tag eq "a") {
                        my $atag = {};
			$atag->{'href'}="";
			$atag->{'rel'}="";
                        if (defined($attr->{"href"}))  {
                                $atag->{'href'} = $attr->{"href"};
			}
                        if (defined($attr->{"rel"})) {
                        	$atag->{'rel'} = $attr->{"rel"};
                        }
			eval "push \@$tableau,\$atag";
		}
}

sub URL_GetBase {
	my ($url) = @_;
        my $base = url($url)->scheme."://".url($url)->host;
        if (url($url)->port != "80") { $base = $base.":".url($url)->port }
        $base = $base."/";
        return $base;
}


sub GiveMeNextSourceToAnalyze
{

        my $stomp = Net::Stomp->new( { hostname => $activemq_server, port => '61613'} );

        my $frame_connect = Net::Stomp::Frame->new( {
                                                command => "CONNECT",
                                                headers => {
                                                        'client-id' => $daemonName
                                                }});
        $stomp->send_frame($frame_connect);
        $frame_connect = $stomp->receive_frame;

        $stomp->subscribe(
        {   destination             => $activemq_source,
                'ack'                   => 'client',
                'activemq.prefetchSize' => 1,
                'activemq.subscriptionName' => $daemonName,
                'activemq.subcriptionName' =>  $daemonName
        });

        while($stomp->can_read({ timeout => '1' }) eq 1) {
		my $link;
		my  $linkurl;
		my  $linkurls;

    		my $frame = $stomp->receive_frame;
    		my $source= decode_utf8($frame->body); # do something here
		my $url   = $frame->headers->{"correlation-id"};

		logEntry("analyse de $url");
		@taboflinks=();
		my $parserhtml = HTML::Parser->new();
           	$parserhtml->handler(start =>\&parseur_start,"tagname,attr,'taboflinks'");
           	$parserhtml->parse($source);

		foreach $link (@taboflinks) {
                	$linkurl  = url($link->{'href'}, URL_GetBase($url))->abs;   # absolute url
			$linkurls = $linkurl->as_string;                            # 
			$linkurls =~ s/(^http[^\#]*)\#.*/$1/;                       # eliminate anchor (after "#")
			$link->{'href'}=$linkurls;
		}
	
		my $json = to_json(\@taboflinks);
		# you can retrieve the array with  : my @tab = @{decode_json($json)};
		my $frame_producer = Net::Stomp::Frame->new( {
                		body    => $json,
                		command => "SEND",
                		headers => {
                        		"destination"    => $activemq_links,
					"correlation-id" => $url,
					"persistent"     => 'true'

                			}
                });
                $stomp->send_frame($frame_producer);

 		$stomp->ack( { frame => $frame } );
	}
  	$stomp->disconnect;

  return;
}
 
#################################################################################################
# MAIN PROGRAM
#################################################################################################
	my $dieNow        = 0;                                     # used for "infinte loop" construct - allows daemon mode to gracefully exit
	my $sleepMainLoop = 10;                                    # number of seconds to wait between "do something" execution after queue is clear
	 
	# daemonize
	use POSIX qw(setsid);
	chdir '/';
	umask 0;
	open STDIN,  '/dev/null'   or die "Can't read /dev/null: $!";
	open STDOUT, '>>/dev/null' or die "Can't write to /dev/null: $!";
	open STDERR, '>>/dev/null' or die "Can't write to /dev/null: $!";
	 
	# dissociate this process from the controlling terminal that started it and stop being part
	# of whatever process group this process was a part of.
	POSIX::setsid() or die "Can't start a new session.";
	 
	# callback signal handler for signals.
	$SIG{INT} = $SIG{TERM} = $SIG{HUP} = \&signalHandler;
	$SIG{PIPE} = 'ignore';
	 
	# create pid file in /var/run/
	my $pidfile = File::Pid->new( { file => $pidFile, } );
	 
	$pidfile->write or die "Can't write PID file, /dev/null: $!";
	 
	# turn on logging
	if ($logging) {
	    open LOG, ">>$logFile";
	    select((select(LOG), $|=1)[0]); # make the log file "hot" - turn off buffering
	}   

   logEntry("daemon started");

    # Enter loop to do work
    # "infinite" loop where some useful process happens
    until ($dieNow) {
       GiveMeNextSourceToAnalyze(); 
       sleep($sleepMainLoop);
    }

exit;


# add a line to the log file
sub logEntry {
    my ($logText) = @_;
    my ( $sec, $min, $hour, $mday, $mon, $year, $wday, $yday, $isdst ) = localtime(time);
    my $dateTime = sprintf "%4d-%02d-%02d %02d:%02d:%02d", $year + 1900, $mon + 1, $mday, $hour, $min, $sec;
    if ($logging) {
        print LOG "$dateTime|$daemonName|$logText\n";
    }
}
 
# catch signals and end the program if one is caught.
sub signalHandler {
    $dieNow = 1;    # this will cause the "infinite loop" to exit
}
 
# do this stuff when exit() is called.
END {

    logEntry("daemon stopped");
    if ($logging) { close LOG }
    $pidfile->remove if defined $pidfile;
}



crawler.pl
#!/usr/bin/perl 
#
#
use strict;
use warnings;
use Net::Stomp;
use LWP::UserAgent;
use Data::Dump qw(dump);
use JSON;
use Encode;
use POSIX;
use File::Pid;
use AppConfig;
use URI::URL;

my $daemonName = "MQcrawler";

#=======================================
# Lecture du fichier de config
#=======================================
my $config = AppConfig->new();
   $config->define("patterns",      {ARGCOUNT => AppConfig::ARGCOUNT_LIST});
   $config->define("redis_server",  {ARGCOUNT => AppConfig::ARGCOUNT_ONE});
   $config->define("redis_set_urls",  {ARGCOUNT => AppConfig::ARGCOUNT_ONE});
   $config->define("redis_set_urls_crawled",  {ARGCOUNT => AppConfig::ARGCOUNT_ONE});
   $config->define("activemq_server",  {ARGCOUNT => AppConfig::ARGCOUNT_ONE});
   $config->define("activemq_source",  {ARGCOUNT => AppConfig::ARGCOUNT_ONE});
   $config->define("activemq_links",  {ARGCOUNT => AppConfig::ARGCOUNT_ONE});
   $config->define("activemq_crawl",  {ARGCOUNT => AppConfig::ARGCOUNT_ONE});
   $config->define("logging",       {ARGCOUNT => AppConfig::ARGCOUNT_ONE});
   $config->define("logpath",       {ARGCOUNT => AppConfig::ARGCOUNT_ONE});
   $config->define("pidpath",       {ARGCOUNT => AppConfig::ARGCOUNT_ONE});
   $config->file('YSpider.conf');

   my @patterns_to_crawl   = ();
   my $patterns = $config->patterns;
   my $i;
   foreach $i (@$patterns) {
                push( @patterns_to_crawl, $i);
   }
   my $redis_server     = $config->redis_server;
   my $activemq_server  = $config->activemq_server;
   my $activemq_source  = $config->activemq_source;
   my $activemq_links   = $config->activemq_links;
   my $activemq_crawl   = $config->activemq_crawl;
   my $logging          = $config->logging();                           # 1= logging is on
   my $logFilePath      = $config->logpath();                           # log file path
   my $pidFilePath      = $config->pidpath();                           # PID file path
      $logFilePath      =~ s/([^\/])$/$1\//; # add slash / if not specified in config 
      $pidFilePath      =~ s/([^\/])$/$1\//; # add slash / if not specified in config
   my $logFile          = $logFilePath . $daemonName . ".log";
   my $pidFile          = $pidFilePath . $daemonName . ".pid";

# global variables
my @redirects=();
my @headers=();

sub GiveMeNextURLToCrawl
{
	my $pid;
	my $ichild=0;
	for (1..10){
		$ichild ++;
		sleep(0.1);
   		$pid = fork();
   		last unless defined $pid;  # Too many processes already?
   		unless($pid){
		my $stomp = Net::Stomp->new( { hostname => $activemq_server, port => '61613' } );
		eval {
			# Child code here
        		$stomp->connect();
        		$stomp->subscribe({destination  => $activemq_crawl, 'ack' => 'client', 'activemq.prefetchSize' => 1 });
        		if ($stomp->can_read({ timeout => '1' }) eq 1) {
	    			my $frame = $stomp->receive_frame;
    				my $url= $frame->body; 
				logEntry("(child $ichild) : crawl de $url");

				my $ua = LWP::UserAgent->new;
        			$ua->timeout(10);
				@redirects = ();
				$ua->add_handler("response_redirect", sub {  my($response, $ua, $h) = @_;
                                                                                 my $url      = $response->request->uri;
                                                                                 my $code     = $response->code;
                                                                                 my $location = $response->header("location");
                                                                                 if (($code eq "301")||($code eq "302")) {
                                                                                        my $redirect = {};
                                                                                        $redirect->{"url"} = $url->as_string;
                                                                                        $location=url($location, URL_GetBase($url))->abs;
                                                                                        $redirect->{"location"} = $location->as_string;
                                                                                        $redirect->{"code"} = $code;
                                                                                        push (@redirects, $redirect);
                                                                                 };
                                                                            return });

				my $res = $ua->get($url);
					my $source = "";
					if ( $res->header("content-type") =~  /text\/htm/i) {
					#if ( $res->header("content-type") =~  /.*text\/htm.*/i) {
						$source = $res->decoded_content;
					}
					@headers=();
					my $h=$res->headers;
					$h->remove_header("Link");  $h->remove_header("X-Meta-Description");  $h->remove_header("X-Meta-Keywords");  $h->remove_header("X-Meta-Robots");
					$h->remove_header("Title");  $h->remove_header("Client-Response-Num");
					$h->scan( sub {  my ($a,$b) = @_;  my $header ={};  $header->{$a}=$b;  push @headers, $header;  } );

					my $frameproducer = Net::Stomp::Frame->new( {
  						body    => encode_utf8($source),
  						command => "SEND",
  						headers => {  "mqcrawler.initial_url"   => $frame->body,
							"mqcrawler.final_url"           => $res->request->uri_canonical->as_string,
               						"mqcrawler.redirects" 		=> to_json(\@redirects),
               						"mqcrawler.http_headers" 	=> to_json(\@headers),
               						"mqcrawler.status_code" 	=> $res->code,
               						"correlation-id" 		=> $frame->body,
               						"destination"    		=> $activemq_source,
							"persistent"     		=> 'true'
	     						}
						});
					$stomp->send_frame($frameproducer);
    					$stomp->ack( { frame => $frame } );
			}
		};
		logEntry($@) if $@;
		eval {  $stomp->disconnect; };
        	exit;
		}
	}
	# wait() for kids
	while(($pid = wait()) > 0){  sleep(0.2); }
	return;
}
 
#################################################################################################
# MAIN PROGRAM
#################################################################################################
        my $dieNow        = 0;                                     # used for "infinte loop" construct - allows daemon mode to gracefully exit
        my $sleepMainLoop = 1;                                    # number of seconds to wait between "do something" execution after queue is clear

        # daemonize
        use POSIX qw(setsid);
        chdir '/';
        umask 0;
        open STDIN,  '/dev/null'   or die "Can't read /dev/null: $!";
        open STDOUT, '>>/dev/null' or die "Can't write to /dev/null: $!";
        open STDERR, '>>/dev/null' or die "Can't write to /dev/null: $!";

        # dissociate this process from the controlling terminal that started it and stop being part
        # of whatever process group this process was a part of.
        POSIX::setsid() or die "Can't start a new session.";

        # callback signal handler for signals.
        $SIG{INT} = $SIG{TERM} = $SIG{HUP} = \&signalHandler;
        $SIG{PIPE} = 'ignore';

        # create pid file in /var/run/
        my $pidfile = File::Pid->new( { file => $pidFile, } );
        $pidfile->write or die "Can't write PID file, /dev/null: $!";

        # turn on logging
        if ($logging) {
            open LOG, ">>$logFile";
            select((select(LOG), $|=1)[0]); # make the log file "hot" - turn off buffering
        }

    logEntry("daemon started");

    # Enter loop to do work
    # "infinite" loop where some useful process happens
    until ($dieNow) {
	GiveMeNextURLToCrawl();
        sleep($sleepMainLoop);
    }

    logEntry("daemon stopped");
    exit;

# add a line to the log file
sub logEntry {
    my ($logText) = @_;
    my ( $sec, $min, $hour, $mday, $mon, $year, $wday, $yday, $isdst ) = localtime(time);
    my $dateTime = sprintf "%4d-%02d-%02d %02d:%02d:%02d", $year + 1900, $mon + 1, $mday, $hour, $min, $sec;
    if ($logging) {
        print LOG "$dateTime|$daemonName|$logText\n";
    }
}

# catch signals and end the program if one is caught.
sub signalHandler {
    $dieNow = 1;    # this will cause the "infinite loop" to exit
}

# do this stuff when exit() is called.
END {
    if ($logging) { close LOG }
    $pidfile->remove if defined $pidfile;
}
                            
sub URL_GetBase {
my ($url) = @_;
        my $base = url($url)->scheme."://".url($url)->host;
        if (url($url)->port != "80") { $base = $base.":".url($url)->port }
        $base = $base."/";
        return $base;
}

 
